core/src/main/scala/kafka/log/Log.scala

```diff
@@ -565,8 +565,8 @@ class Log(val dir: File,
   def deleteOldSegments(predicate: LogSegment => Boolean): Int = {
     // find any segments that match the user-supplied predicate UNLESS it is the final segment
     // and it is empty (since we would just end up re-creating it
-    val lastSegment = activeSegment
-    val deletable = logSegments.takeWhile(s => predicate(s) && (s.baseOffset != lastSegment.baseOffset || s.size > 0))
+    val lastEntry = segments.lastEntry
+    val deletable = logSegments.takeWhile(s => predicate(s) && (lastEntry != null && s.baseOffset != lastEntry.getValue.baseOffset || s.size > 0))
```

Comment:
You can do the following, right?

```scala
val deletable = lastEntry != null && logSegments.takeWhile(s => predicate(s) && (s.baseOffset != lastSegment.baseOffset || s.size > 0))
```

Also, I think you'd want to extract `logSegments.takeWhile(...)` into an inner method because you need to call it again after the lock is taken in case it's changed due to concurrent `delete` call, right?

Also, we'd want to document that `deleteOldSegments` can be called concurrently with `delete`. I was wondering if this is by design. We abort `LogCleaner` for a given partition before calling `Log.delete`, but we make no effort to deal with scheduled `LogManager.cleanupLogs`.

It would be good to get @junrao's input on whether this is the right approach to take.